src/deploy/functions/release/sourceTokenScraper.ts

```diff
@@ -8,40 +13,54 @@ import { Timer } from "./timer";
  */
 export class SourceTokenScraper {
   private tokenValidDurationMs; // in ms
-  private tokenRefreshRequired = true;
-  private tokenRefreshTimer: Timer | undefined;
   private resolve!: (token?: string) => void;
   private promise: Promise<string | undefined>;
+  private expiration: bigint | undefined;
+  private tokenState: TokenState;
 
   constructor(validDurationMs = 1500000) {
     this.tokenValidDurationMs = validDurationMs;
     this.promise = new Promise((resolve) => (this.resolve = resolve));
+    this.tokenState = TokenState.NONE;
   }
 
-  // Token Promise will return undefined for the first caller
-  // (because we presume it's this function's source token we'll scrape)
-  // and then returns the promise generated from the first function's onCall
-  tokenPromise(): Promise<string | undefined> {
-    if (this.tokenRefreshRequired) {
-      this.tokenRefreshRequired = false;
-      this.tokenRefreshTimer = new Timer(this.tokenValidDurationMs);
-      return Promise.resolve(undefined);
+  async getToken(): Promise<string | undefined> {
+    if (this.tokenState === TokenState.NONE) {
+      this.tokenState = TokenState.INVALID;
+      return undefined;
+    } else if (this.tokenState === TokenState.INVALID) {
+      return this.promise; // wait until we get a source token
+    } else if (this.tokenState === TokenState.VALID) {
+      if (this.checkTokenExpired()) {
+        this.tokenState = TokenState.INVALID;
+        this.promise = new Promise((resolve) => (this.resolve = resolve));
+        return undefined;
+      }
+      return this.promise;
     }
-    return this.promise;
+  }
+
+  checkTokenExpired(): boolean {
+    if (this.expiration === undefined) {
+      throw new Error();
```

Comment:
This error is user visible - we should try our best to describe the situation and also call out what users should do when they see the issue.

Reply:
'throw new FirebaseError(
        "Your deployment is checking the expiration of a source token that has not yet been polled. " +
          "Hitting this case should never happen and should be considered a bug. " +
          "Please file an issue at https://github.com/firebase/firebase-tools/issues"
      );'